src/Numerics/DGMethods/DGFVModel_kernels.jl

```diff
@@ -0,0 +1,800 @@
+                sgeo[_n3, n, faces[1], eH + eV_up],
+            )
+
+            # Reconstruction for eV_dn was computed in last time through the
+            # loop, so we need to store the upper reconstructed values to
+            # compute flux for this face
+            local_state_face_prognostic_neighbor .=
+                local_state_face_prognostic[2]
+
+            # Next data we need to load (assume periodic, mod1, for now  will
+            # mask out below as needed for boundary conditions)
+            eV_load = mod1(eV_up + stencil_width, nvertelem)
+
+            # get element number
+            e_load = eH + eV_load
+
+            # Load the next cell into the end of the element arrays
+            load_data!(
+                local_state_prognostic[stencil_diameter],
+                local_state_auxiliary[stencil_diameter],
+                local_state_gradient_flux[stencil_diameter],
+                e_load,
+            )
+
+            # Get local volume mass matrix inverse
+            local_cell_weights[stencil_diameter] = vgeo[n, _M, e_load]
+            vMI[2] = 1 / local_cell_weights[stencil_center]
+
+            # tranform the prognostic data to primitive data
+            prognostic_to_primitive!(
+                balance_law,
+                local_state_primitive[stencil_diameter],
+                local_state_prognostic[stencil_diameter],
+                local_state_auxiliary[stencil_diameter],
+            )
+
+            # Do the reconstruction! for this cell and compute the values at the
+            # bottom (1) and top (2) faces of element `eV_up`
+            if periodicstack ||
+               stencil_width < eV_up < nvertelem - stencil_width + 1
+                # If we are in the interior or periodic just use the
+                # reconstruction
+                rng1, rng2 = stencil_center .+ (-stencil_width, stencil_width)
+                rng = SUnitRange(rng1, rng2)
+                reconstruction!(
+                    local_state_face_primitive[1],
+                    local_state_face_primitive[2],
+                    local_state_primitive[rng],
+                    local_cell_weights[rng],
+                )
+            elseif eV_up <= stencil_width
+                # Bottom of the element stack requires reconstruct using a
+                # subset of the elements
+                # Values around stencil center that we need for this
+                # reconstruction
+                Base.Cartesian.@nif 4 w -> (eV_up == w) w -> begin
+                    rng1, rng2 = stencil_center .+ (1 - w, w - 1)
+                    rng = SUnitRange(rng1, rng2)
+                    reconstruction!(
+                        local_state_face_primitive[1],
+                        local_state_face_primitive[2],
+                        local_state_primitive[rng],
+                        local_cell_weights[rng],
+                    )
+                end w -> throw(BoundsError(local_state_primitive, w))
```

Comment:
Is it rng = SUnitRange(stencil_center - eV_up+1, stencil_center + eV_up-1) ? 
Why do we need to throw error ?

Reply:
If this was a CPU only code, we could use `eV_up`, but on the GPU we need to generate a serious of `if...elseif....else` statements for each possible value of `eV_up` that could be encountered. So `w` in this code block takes the place of `eV_up` which will be set directly by the [`Base.Cartesian.@nif'](https://docs.julialang.org/en/v1/devdocs/cartesian/#Base.Cartesian.@nif) macro; the trouble with the GPU is though we have "dynamic syntax" the code still needs to be analyzed statically.

The `error` should never get hit since the `@assert stencil_width < 4` at the top of the code will prevent it, but I was just being defensive here. The error will only getting thrown by the `else` statement of the macro. (What's written should work for upto WENO5, if we go to WENO7 we would need to change the `4` to a `5`.

Reply:
> I think this PR is fully tested, (the spherical test has 1 element < stencil width in vertical so the FVLinear reconstruction might give slightly different results)

I increased the number of elements to 3 in the spherical test (this was necessary to get your original linear reconstruction kernel to run on the test). The kernel @thomasgibson and I wrote, should work with 1 element in the vertical, and we could decrease it again if you wanted (though the numbers would need to be updated).